File: src/main/java/org/example/kalkulationsprogramm/repository/AusgangsGeschaeftsDokumentRepository.java
package org.example.kalkulationsprogramm.repository;

import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokument;
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentTyp;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface AusgangsGeschaeftsDokumentRepository extends JpaRepository<AusgangsGeschaeftsDokument, Long> {

    /**
     * Findet alle Dokumente für ein Projekt, sortiert nach Datum absteigend
     */
    List<AusgangsGeschaeftsDokument> findByProjektIdOrderByDatumDesc(Long projektId);

    /**
     * Findet alle Dokumente für ein Anfrage, sortiert nach Datum absteigend
     */
    List<AusgangsGeschaeftsDokument> findByAnfrageIdOrderByDatumDesc(Long anfrageId);

    /**
     * Findet alle Dokumente für einen Kunden, sortiert nach Datum absteigend
     */
    List<AusgangsGeschaeftsDokument> findByKundeIdOrderByDatumDesc(Long kundeId);

    /**
     * Findet Dokument anhand der Dokumentnummer
     */
    Optional<AusgangsGeschaeftsDokument> findByDokumentNummer(String dokumentNummer);

    /**
     * Findet alle Nachfolger-Dokumente eines Vorgängers, sortiert nach Erstellungszeitpunkt
     */
    List<AusgangsGeschaeftsDokument> findByVorgaengerIdOrderByErstelltAmAsc(Long vorgaengerId);

    /**
     * Zählt Abschlagsrechnungen zu einem Vorgänger-Dokument (für Nummerierung)
     */
    @Query("SELECT COUNT(d) FROM AusgangsGeschaeftsDokument d WHERE d.vorgaenger.id = :vorgaengerId AND d.typ = :typ")
    int countByVorgaengerIdAndTyp(Long vorgaengerId, AusgangsGeschaeftsDokumentTyp typ);

    /**
     * Findet das (eindeutige) Dokument eines bestimmten Typs für ein Anfrage.
     * z.B. das ANFRAGE-Dokument, aus dem die Anfragesnummer abgeleitet wird.
     */
    Optional<AusgangsGeschaeftsDokument> findFirstByAnfrageIdAndTyp(Long anfrageId, AusgangsGeschaeftsDokumentTyp typ);

    /**
     * Prüft ob bereits ein Basisdokument (ohne Vorgänger) für ein Projekt existiert.
     */
    boolean existsByProjektIdAndVorgaengerIsNull(Long projektId);

    /**
     * Prüft ob bereits ein Basisdokument (ohne Vorgänger) für ein Anfrage existiert.
     */
    boolean existsByAnfrageIdAndVorgaengerIsNull(Long anfrageId);

    /**
     * Findet alle Dokumente, die über ein Anfrage mit einem Projekt verknüpft sind
     * (anfrage.projekt_id = projektId), aber kein direktes projekt_id gesetzt haben.
     */
    @Query("SELECT d FROM AusgangsGeschaeftsDokument d WHERE d.anfrage.projekt.id = :projektId AND d.projekt IS NULL")
    List<AusgangsGeschaeftsDokument> findByAnfrageProjektIdAndProjektIsNull(Long projektId);

    /**
     * Findet die höchste Dokumentnummer für einen Monat (für neue Nummer)
     */
    @Query("SELECT MAX(d.dokumentNummer) FROM AusgangsGeschaeftsDokument d WHERE d.dokumentNummer LIKE :prefix%")
    Optional<String> findMaxDokumentNummerByPrefix(String prefix);

    /**
     * Findet alle Dokumente in einem Datumsbereich (für Dokumentübersicht).
     */
    List<AusgangsGeschaeftsDokument> findByDatumBetweenOrderByDatumDesc(java.time.LocalDate start, java.time.LocalDate end);

    /**
     * Findet alle Dokumente sortiert nach Datum absteigend (für Dokumentübersicht ohne Datumsfilter).
     */
    List<AusgangsGeschaeftsDokument> findAllByOrderByDatumDesc();

    /**
     * Liefert für eine Liste von Anfrage-IDs ein Mapping
     * [AusgangsGeschaeftsDokument.id, Anfrage.id], damit der DokumentFreigabeService
     * den jüngsten Freigabe-Status pro Anfrage auch für Dokumente aus dem neuen
     * AusgangsGeschaeftsDokument-System finden kann.
     */
    @Query("SELECT d.id, d.anfrage.id FROM AusgangsGeschaeftsDokument d WHERE d.anfrage.id IN :anfrageIds")
    List<Object[]> findIdAnfrageIdMappingByAnfrageIds(@org.springframework.data.repository.query.Param("anfrageIds") List<Long> anfrageIds);

    /**
     * Pendant zu {@link #findIdAnfrageIdMappingByAnfrageIds} für Projekte.
     */
    @Query("SELECT d.id, d.projekt.id FROM AusgangsGeschaeftsDokument d WHERE d.projekt.id IN :projektIds")
    List<Object[]> findIdProjektIdMappingByProjektIds(@org.springframework.data.repository.query.Param("projektIds") List<Long> projektIds);
}

File: src/main/kotlin/org/example/kalkulationsprogramm/repository/AusgangsGeschaeftsDokumentRepository.kt
package org.example.kalkulationsprogramm.repository

import java.time.LocalDate
import java.util.Optional
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokument
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentTyp
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface AusgangsGeschaeftsDokumentRepository : JpaRepository<AusgangsGeschaeftsDokument, Long> {
    fun findByProjektIdOrderByDatumDesc(projektId: Long?): List<AusgangsGeschaeftsDokument>

    fun existsByProjektId(projektId: Long?): Boolean

    fun findByAnfrageIdOrderByDatumDesc(anfrageId: Long?): List<AusgangsGeschaeftsDokument>

    fun findByKundeIdOrderByDatumDesc(kundeId: Long?): List<AusgangsGeschaeftsDokument>

    fun findByDokumentNummer(dokumentNummer: String): Optional<AusgangsGeschaeftsDokument>

    fun findByVorgaengerIdOrderByErstelltAmAsc(vorgaengerId: Long?): List<AusgangsGeschaeftsDokument>

    @Query("SELECT COUNT(d) FROM AusgangsGeschaeftsDokument d WHERE d.vorgaenger.id = :vorgaengerId AND d.typ = :typ")
    fun countByVorgaengerIdAndTyp(vorgaengerId: Long?, typ: AusgangsGeschaeftsDokumentTyp): Int

    fun findFirstByAnfrageIdAndTyp(
        anfrageId: Long?,
        typ: AusgangsGeschaeftsDokumentTyp,
    ): Optional<AusgangsGeschaeftsDokument>

    fun existsByProjektIdAndVorgaengerIsNull(projektId: Long?): Boolean

    fun existsByAnfrageIdAndVorgaengerIsNull(anfrageId: Long?): Boolean

    fun existsByProjektIdAndVorgaengerIsNullAndTyp(
        projektId: Long?,
        typ: AusgangsGeschaeftsDokumentTyp,
    ): Boolean

    fun existsByAnfrageIdAndVorgaengerIsNullAndTyp(
        anfrageId: Long?,
        typ: AusgangsGeschaeftsDokumentTyp,
    ): Boolean

    @Query("SELECT d FROM AusgangsGeschaeftsDokument d WHERE d.anfrage.projekt.id = :projektId AND d.projekt IS NULL")
    fun findByAnfrageProjektIdAndProjektIsNull(projektId: Long?): List<AusgangsGeschaeftsDokument>

    @Query("SELECT MAX(d.dokumentNummer) FROM AusgangsGeschaeftsDokument d WHERE d.dokumentNummer LIKE :prefix%")
    fun findMaxDokumentNummerByPrefix(prefix: String): Optional<String>

    fun findByDatumBetweenOrderByDatumDesc(start: LocalDate, end: LocalDate): List<AusgangsGeschaeftsDokument>

    fun findAllByOrderByDatumDesc(): List<AusgangsGeschaeftsDokument>

    @Query("SELECT d.id, d.anfrage.id FROM AusgangsGeschaeftsDokument d WHERE d.anfrage.id IN :anfrageIds")
    fun findIdAnfrageIdMappingByAnfrageIds(
        @Param("anfrageIds") anfrageIds: List<Long>,
    ): List<Array<Any>>

    @Query("SELECT d.id, d.projekt.id FROM AusgangsGeschaeftsDokument d WHERE d.projekt.id IN :projektIds")
    fun findIdProjektIdMappingByProjektIds(
        @Param("projektIds") projektIds: List<Long>,
    ): List<Array<Any>>

    @Query(
        "SELECT d.rechnungsadresseOverride FROM AusgangsGeschaeftsDokument d " +
            "LEFT JOIN d.projekt p LEFT JOIN d.anfrage a LEFT JOIN a.projekt ap " +
            "WHERE (p.id = :projektId OR ap.id = :projektId) " +
            "AND d.rechnungsadresseOverride IS NOT NULL AND TRIM(d.rechnungsadresseOverride) <> '' " +
            "ORDER BY d.geaendertAm DESC",
    )
    fun findRechnungsadresseOverridesByProjektId(
        @Param("projektId") projektId: Long?,
    ): List<String>

    @Query(
        "SELECT d.rechnungsadresseOverride FROM AusgangsGeschaeftsDokument d " +
            "WHERE d.anfrage.id = :anfrageId " +
            "AND d.rechnungsadresseOverride IS NOT NULL AND TRIM(d.rechnungsadresseOverride) <> '' " +
            "ORDER BY d.geaendertAm DESC",
    )
    fun findRechnungsadresseOverridesByAnfrageId(
        @Param("anfrageId") anfrageId: Long?,
    ): List<String>
}
